lib/types.d.ts

```diff
@@ -0,0 +1,32 @@
+type MergeBy<TObject, TOtherObject> = Omit<TObject, keyof TOtherObject> & TOtherObject;
+
+type DeepRecord<TKey extends string | number | symbol, TValue> = {[key: string]: TValue | DeepRecord<TKey, TValue>};
+
+type DeepKeyOf<TObject> = TObject extends object
+    ? {
+          [TKey in keyof TObject & (string | number)]: TObject[TKey] extends Record<string, unknown> ? `${TKey}.${DeepKeyOf<TObject[TKey]>}` | TKey : TKey;
+      }[keyof TObject & (string | number)]
+    : TObject;
+
+type TypeOptions = MergeBy<
+    {
+        keys: string;
+        collectionKeys: string;
+        values: Record<string, unknown>;
+    },
+    CustomTypeOptions
+>;
+
+interface CustomTypeOptions {}
+
+type Key = TypeOptions['keys'];
+type CollectionKey = TypeOptions['collectionKeys'];
+type Value = TypeOptions['values'];
```

Comment:
NAB:

1. Shouldn't value type be in plural? `Values` instead of `Value`.
2. `Key | CollectionKey` is used in many placed we could extract it to a type and reuse.
3. Change `Key` to some more descriptive name? My idea is `NonCollectionKey` but maybe it's too much haha

```suggestion
type NonCollectionKey = TypeOptions['keys'];
type CollectionKey = TypeOptions['collectionKeys'];
type Key = NonCollectionKey | CollectionKey; // this could be reused in various functions
type Values = TypeOptions['values'];
```

Reply:
1. I think we want types to be singular, Value represents the type of any possible value right?
2. Sounds good, maybe we can do `type OnyxKey = Key | CollectionKey`?

Reply:
1. I agree with @neil-marcellini, should be `Value` the same way `Key` is.
2. I like `OnyxKey` idea, wdyt @blazejkustra ?

Reply:
> Sounds good, maybe we can do type OnyxKey = Key | CollectionKey?

I like the idea. Then we don't have to constantly do `Key | CollectionKey` throughout type definitions

> I agree with @neil-marcellini, should be Value the same way Key is.

Key is an union type, but `Values` is a mapping between key and value pairs. I think this should be plural or this should be something like `OnyxData` or `KeyValueMapping`

Reply:
> 2. I like OnyxKey idea, wdyt @blazejkustra ?

Sounds good!

> Key is an union type, but Values is a mapping between key and value pairs. I think this should be plural or this should be something like OnyxData or KeyValueMapping

+1

Reply:
Ah I see about it being plural now. I like `KeyValueMapping`.

Reply:
Will adjust!